File: tools/book_workshops.py
"""Scaffolded close reading and small source-bounded experiments."""
from html import escape as E

def worked(l):
 d=l.get('pedagogy')
 if not d:return ''
 w=d['worked']
 return '<section class="book-worked"><p class="eyebrow">Смотрим, как строится объяснение</p><h2>'+E(w['question'])+'</h2><ol>'+''.join('<li>'+E(s)+'</li>' for s in w['steps'])+'</ol><p class="fine">Теперь закройте этот пример и попробуйте следующий случай самостоятельно.</p></section>'

def hints(l):
 if 'pedagogy' not in l:return ''
 return '<div class="book-hints"><h3>Если понадобилась помощь</h3>'+''.join('<details><summary>Подсказка '+str(i+1)+'</summary><p>'+E(t)+'</p></details>' for i,t in enumerate(l['pedagogy']['hints']))+'</div>'

def rubric(l):
 if 'pedagogy' not in l:return ''
 return '<div class="book-rubric"><h3>Проверьте свой ответ</h3><p>Для каждого пункта найдите фразу в своей записи. Если её нет — дополните ответ. Автоматическая проверка открытого текста не выполняется.</p><ul>'+''.join('<li>'+E(x)+'</li>' for x in l['pedagogy']['rubric'])+'</ul></div>'

def interactive(k):
 def btn(t,a):return '<button type="button" class="button secondary" disabled data-book-enhanced '+a+'>'+E(t)+'</button>'
 if k['type']=='kav':
  return '<figure class="kav-figure"><svg viewBox="0 0 440 310" role="img" aria-labelledby="kav-title kav-desc"><title id="kav-title">Модель связи: ограниченный кав</title><desc id="kav-desc">Окружность условно отмечает границу свободной области; линия идёт от верхней границы и не достигает нижней. Рисунок показывает отношения, а не размеры Эйн Соф.</desc><circle cx="220" cy="156" r="128" fill="#f7f9f5" stroke="#a98a50" stroke-width="4"/><path data-kav-line d="M220 28V231" stroke="#244b54" stroke-width="5"/><path data-kav-bottom d="M220 231V284" stroke="#244b54" stroke-width="5" opacity="0"/><circle cx="220" cy="28" r="6" fill="#244b54"/><text x="240" y="92">Начало</text><text x="240" y="229">Конец</text></svg><figcaption>Граница области условна. Изменяется только способ связи; масштабы и физические расстояния не показаны.</figcaption></figure><div class="sy-actions">'+''.join(btn(t,'data-kav="'+v+'" aria-pressed="'+str(v=='single').lower()+'"') for v,t in [('none','Без линии'),('single','Один связанный конец'),('double','Два связанных конца')])+'</div><p class="sy-output" data-kav-output role="status">Один связанный конец: сохраняются поступление и различие начала и конца.</p>'
 if k['type']=='fives':
  return '<div class="sy-actions">'+''.join(btn(x['title'],'data-five="'+str(i)+'" aria-pressed="false"') for i,x in enumerate(k['cards']))+'</div><div class="book-experiment-output" data-five-output role="status"><p>Выберите пятёрку после собственного предположения.</p></div><details class="sy-solution"><summary>Все четыре описания для сверки</summary>'+''.join('<h3>'+E(x['title'])+'</h3><p>'+E(x['pair']+' · '+x['name'])+'</p><p>'+E(x['image'])+'</p><p>'+E(x['meaning'])+'</p>' for x in k['cards'])+'</details>'
 if k['type']=='dignities':
  choices=[('wands','Жезлы'),('cups','Кубки'),('swords','Мечи'),('pentacles','Пентакли')]
  def select(pos,label):return '<label for="dignity-'+pos+'">'+label+'</label><select id="dignity-'+pos+'" data-dignity-'+pos+' disabled data-book-enhanced>'+''.join('<option value="'+v+'">'+t+'</option>' for v,t in choices)+'</select>'
  return '<div class="dignity-controls">'+select('left','Масть слева')+'<p><strong>В центре: Пятёрка Жезлов</strong></p>'+select('right','Масть справа')+'</div>'+btn('Разобрать соседство','data-dignity-check')+'<p class="sy-output" data-dignity-output role="status">Выберите соседей, сформулируйте ожидание и нажмите кнопку.</p><p class="fine">Числовой балл не вычисляется: источник не задаёт полной шкалы смешанных сочетаний.</p>'
 raise ValueError(k['type'])

File: assets/course/book-workshops.js
'use strict';
(()=>{
 const node=document.querySelector('#book-module-data');if(!node)return;
 const d=JSON.parse(node.textContent),$=s=>document.querySelector(s),$$=s=>Array.from(document.querySelectorAll(s));
 const kav={none:'Без линии: свободная область обозначена, но способ поступления света к мирам не объяснён.',single:'Один связанный конец: поступление сохраняется, а выделенное начало позволяет различать верх и низ. Это условие выбранного аргумента автора.',double:'Два связанных конца: при одинаковом поступлении оба оказываются равноценными началами. Прежнее основание различать верх и низ исчезает; это контрпример самого текста.'};
 $$('[data-kav]').forEach(b=>b.addEventListener('click',()=>{
  const mode=b.dataset.kav;
  $$('[data-kav]').forEach(x=>x.setAttribute('aria-pressed',String(x===b)));
  $('[data-kav-line]').setAttribute('opacity',mode==='none'?'0':'1');
  $('[data-kav-bottom]').setAttribute('opacity',mode==='double'?'1':'0');
  $('[data-kav-output]').textContent=kav[mode];$('#kav-title').textContent='Модель связи: '+b.textContent;$('#kav-desc').textContent=kav[mode];
 }));
 $$('[data-five]').forEach(b=>b.addEventListener('click',()=>{
  $$('[data-five]').forEach(x=>x.setAttribute('aria-pressed',String(x===b)));
  const c=d.lab.cards[Number(b.dataset.five)],out=$('[data-five-output]');out.replaceChildren();
  for(const [tag,text] of [['h3',c.title],['p','Общее: пятёрка · Гвура.'],['p',c.pair+' · '+c.name],['p',c.image],['p',c.meaning]]){const el=document.createElement(tag);el.textContent=text;out.append(el);}
 }));
 const check=$('[data-dignity-check]'),out=$('[data-dignity-output]');
 if(check){check.addEventListener('click',()=>out.textContent=BookWorkshops.dignity('wands',$('[data-dignity-left]').value,$('[data-dignity-right]').value).text);
  $$('[data-dignity-left],[data-dignity-right]').forEach(s=>s.addEventListener('change',()=>out.textContent='Соседство изменено. Сначала предположите результат, затем нажмите «Разобрать соседство».'));
 }
})();
